prov/efa/src/efa_hmem.c

```diff
@@ -31,9 +31,11 @@ static size_t efa_max_eager_msg_size_with_largest_header(struct efa_domain *efa_
 static int efa_domain_hmem_info_init_protocol_thresholds(struct efa_domain *efa_domain, enum fi_hmem_iface iface)
 {
 	struct efa_hmem_info *info = &efa_domain->hmem_info[iface];
+	size_t tmp_value;
 
-	/* Fall back to FI_HMEM_SYSTEM initialization logic when p2p is unavailable */
-	if (!info->p2p_supported_by_device)
+	/* Fall back to FI_HMEM_SYSTEM initialization logic when p2p is
+	 * unavailable */
+	if (ofi_hmem_p2p_disabled() || !info->p2p_supported_by_device)
 		iface = FI_HMEM_SYSTEM;
```

Comment:
We only need to check `p2p_supported_by_device` and return 0 if it's false.

Reply:
> Fall back to FI_HMEM_SYSTEM initialization logic when p2p is unavailable

Is this not true any more?

Reply:
This function should only initialize 1 iface at a time. I don't see the value of the fallback logic.

Reply:
The function still initializes the single iface (`efa_domain->hmem_info[iface]` assigned to `info`); The `iface` parameter is simply overwritten here after for convenience to trigger a different iface's logic.

Reply:
`efa_domain_hmem_info_init_protocol_thresholds` is called for each iface - keeping this logic means FI_HMEM_SYSTEM could be initialized more than once. I don't think that is necessary.

Reply:
I somehow agree with wenduo, I think we should only initialize system iface once. For non-system iface and p2p disabled/not supported, we can just return instead of doing things for system iface again

Reply:
No. The `iface` overwrite has no effect on which iface is getting initialized since the `info` initialization at the top of the function is independent of it. `FI_HMEM_SYSTEM` would not be initialized again if e.g. the function is called with `FI_HMEM_CUDA` and P2P is unavailable.


```c
// Fetch intended hmem_info entry; i.e. the "iface" in this context
struct efa_hmem_info *info = &efa_domain->hmem_info[iface];

...

// Overwrite the iface var for the purpose of dispatching to FI_HMEM_SYSTEM
// logic in the switch-case construct below
if (ofi_hmem_p2p_disabled() || !info->p2p_supported_by_device)
	iface = FI_HMEM_SYSTEM;
```

Reply:
I see, you are right, it is actually initializing cuda's iface while using system's threshold.

Reply:
I see. The confusion originates from the comment above
> // Overwrite the iface var for the purpose of dispatching to FI_HMEM_SYSTEM
// logic in the switch-case construct below

A more concise and less confusing logic will be
```
struct efa_hmem_info *info = &efa_domain->hmem_info[iface];

// Set default values
info->runt_size = 0;
info->max_medium_msg_size = EFA_DEFAULT_INTER_MAX_MEDIUM_MESSAGE_SIZE;
info->min_read_msg_size = EFA_DEFAULT_INTER_MIN_READ_MESSAGE_SIZE;
info->min_read_write_size = EFA_DEFAULT_INTER_MIN_READ_WRITE_SIZE;
fi_param_get_size_t(&efa_prov, "runt_size", &info->runt_size);
fi_param_get_size_t(&efa_prov, "inter_max_medium_message_size", &info->max_medium_msg_size);
		fi_param_get_size_t(&efa_prov, "inter_min_read_message_size", &info->min_read_msg_size);
fi_param_get_size_t(&efa_prov, "inter_min_read_write_size", &info->min_read_write_size);

if (!info->p2p_supported_by_device) return;

switch (iface) {
case ...:
        ... // override params and check illegal parameters
}

```